docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,283 @@
+
+Local and remote attestation is a TEE-specific feature. Specifying LOCAL or REMOTE flag when generating evidence can be confusing to many attesters. These flags were deprecated.
+
+User Experience
+---------------
+
+The user experience is improved by the ability of querying what the registered attestation verifiers can verify and attesters can attest. This is a typical attestation scenario:
+
+- Verifier application
+    1. Registers appropriate verifier plugins (oe_register_verifier).
+    2. Gets the format IDs of registered verifier plugins (oe_get_registered_verifier_format_ids).
+    3. Sends the format IDs to the attester application. These are the evidence formats that the verifier application is capable of verifying.
+- Attester application
+    1. Receives format IDs from verifier application.
+    2. Registers appropriate attester plugins for the platform so the evidence format selection in the next step can succeed (oe_register_attester).
+    3. Selects a format from the received format IDs (oe_select_attester_format).
+    4. Generates an evidence (oe_get_evidence) of the selected format and sends it to the verifier application.
+- Verifier application
+    1. Verifies the received evidence (oe_verify_evidence).
+
+Specification
+-------------
+
+## Differences between V2 and V3 APIs.
+
+Most of the APIs to attestation plugins are the same between the versions. The APIs to query and select registered attesters and verifiers are added. Each built-in SGX plugin was split into 2 plugins.  The additions and differences are highlighted below.
+
+## Changes
+
+### New APIs to query and select registered attesters and verifiers
+
+```C
+/**
+ * oe_get_registered_attester_format_ids
+ *
+ * Get the format IDs of all registered attesters.
+ *
+ * @param[out] format_ids The list of the format ids of
+ * the registered attesters.
+ * @param[out] format_ids_length The length of the format ids list.
+ * @retval OE_OK on success.
+ * @retval Otherwise, returns the error code the plugin's function.
+ */
+oe_result_t oe_get_registered_attester_format_ids(
+    oe_uuid_t** format_ids,
+    size_t* format_ids_length);
+
+/**
+ * oe_get_registered_verifier_format_ids
+ *
+ * Get the format IDs of all registered verifiers.
+ *
+ * @param[out] format_ids The list of the format ids of
+ * the registered verifiers.
+ * @param[out] format_ids_length The length of the format ids list.
+ * @retval OE_OK on success.
+ * @retval Otherwise, returns the error code the plugin's function.
+ */
+oe_result_t oe_get_registered_verifier_format_ids(
+    oe_uuid_t** format_ids,
+    size_t* format_ids_length);
+
+/**
+ * oe_free_format_ids
+ *
+ * Frees the format ids.
+ *
+ * @param[in] format_ids The list of the format ids.
+ * @retval OE_OK on success.
+ */
+oe_result_t oe_free_format_ids(oe_uuid_t* format_ids);
+
+/**
+ * oe_select_attester_format
+ *
+ * From the list, select an evidence format that the registered attesters
```

Comment:
The input format ID list is sorted by preference by the verifier. The first one in the list is the favorite one. oe_select_attester_format() simply compares and match the first one that it can generate from the list. 

Attester cannot specify its preference. Since attester plugins may come from different vendors, comparing their efficiencies will be unfair. On the other hand, the attester application may choose not to call oe_select_attester_format() and pick its own favorite evidence format.

I will clarify in the comment.

Reply:
Added more comments to oe_select_attester_format.